src/components/tl/ucp/allgatherv/allgatherv_ring.c

```diff
@@ -0,0 +1,94 @@
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "allgatherv.h"
+#include "core/ucc_progress_queue.h"
+#include "utils/ucc_math.h"
+#include "tl_ucp_sendrecv.h"
+
+ucc_status_t ucc_tl_ucp_allgatherv_ring_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task     = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team     = task->team;
+    ucc_rank_t         grank    = team->rank;
+    ucc_rank_t         gsize    = team->size;
+    ptrdiff_t          rbuf     = (ptrdiff_t)task->args.dst.info_v.buffer;
+    ucc_memory_type_t  rmem     = task->args.dst.info_v.mem_type;
+    ucc_datatype_t     rdt_size = ucc_dt_size(task->args.dst.info_v.datatype);
+    ucc_rank_t         sendto   = (grank + 1) % gsize;
+    ucc_rank_t         recvfrom = (grank - 1 + gsize) % gsize;
+    ucc_rank_t         chunk;
+    size_t             data_size, data_displ;
+
+    if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+        return task->super.super.status;
+    }
+    while (task->send_posted < gsize) {
+        chunk = (grank - task->send_posted + 1 + gsize) % gsize;
+        data_displ = ucc_coll_args_get_displacement(&task->args,
+                        task->args.dst.info_v.displacements, chunk) * rdt_size;
+        data_size = ucc_coll_args_get_count(&task->args,
+                        task->args.dst.info_v.counts, chunk) * rdt_size;
+        ucc_tl_ucp_send_nb((void*)(rbuf + data_displ), data_size, rmem, sendto,
+                           team, task);
+        chunk = (grank - task->recv_posted + gsize) % gsize;
+        data_displ = ucc_coll_args_get_displacement(&task->args,
+                        task->args.dst.info_v.displacements, chunk) * rdt_size;
+        data_size = ucc_coll_args_get_count(&task->args,
+                        task->args.dst.info_v.counts, chunk) * rdt_size;
+        ucc_tl_ucp_recv_nb((void*)(rbuf + data_displ), data_size, rmem,
+                           recvfrom, team, task);
+        if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+            return task->super.super.status;
+        }
+    }
+    ucc_assert(UCC_TL_UCP_TASK_P2P_COMPLETE(task));
+    task->super.super.status = UCC_OK;
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_ucp_allgatherv_ring_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task  = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team  = task->team;
+    ptrdiff_t          sbuf  = (ptrdiff_t)task->args.src.info.buffer;
+    ptrdiff_t          rbuf  = (ptrdiff_t)task->args.dst.info_v.buffer;
+    ucc_memory_type_t  smem  = task->args.src.info.mem_type;
+    ucc_memory_type_t  rmem  = task->args.dst.info_v.mem_type;
+    ucc_rank_t         grank = team->rank;
+    size_t             data_size, data_displ, rdt_size;
+    ucc_status_t       status;
+
+    task->super.super.status     = UCC_INPROGRESS;
+    if (!UCC_IS_INPLACE(task->args)) {
+        /* TODO replace local sendrecv with memcpy? */
```

Comment:
why are using send/recv ?

Reply:
because of 
1. memcpy PR wasn't merged yet when I pushed this PR
2. in some use cases memcpy might deadlock e.g if cuda stream or cuda event passed and we wait for that event to complete before starting collective and at the same time user called cuda device synchronize. I think we need memcpy async to support it

Reply:
i see, you are right. What is the right fix then? In general we will maybe need that mc_memcpy_async but in TL/UCP we can be good with send/recv. We should fix the allgather then the same way.